policy/release/olm.rego

```diff
@@ -0,0 +1,99 @@
+#
+# METADATA
+# description: >-
+#   Checks for Operator Lifecycle Manager (OLM) bundles.
+#
+package policy.release.olm
+
+import future.keywords.contains
+import future.keywords.if
+import future.keywords.in
+
+import data.lib
+import data.lib.image
+
+olm_manifestv1 := "operators.operatorframework.io.bundle.manifests.v1"
+
+# METADATA
+# title: Unpinned images in OLM bundle
+# description: >-
+#   Checks the OLM bundle image for the presence of unpinned image references.
+# custom:
+#   short_name: unpinned_references
+#   failure_msg: The %q image reference is not pinned at %s.
+#   solution: >-
+#     Update the OLM bundle replacing the unpinned image reference with pinned image
+#     reference. Pinned image reference contains the image digest.
+#   collections:
+#   - redhat
+#
+deny contains result if {
+	olm_bundle
+	some _, manifest in input.image.manifests
+	manifest.apiVersion == "operators.coreos.com/v1alpha1"
+	manifest.kind == "ClusterServiceVersion"
+
+	some i in all_image_ref(manifest)
+	i.ref.digest == "" # unpinned image references have no digest
+
+	result := lib.result_helper_with_term(rego.metadata.chain(), [image.str(i.ref), i.path], image.str(i.ref))
+}
+
+# Checks if the image is a OPA bundle
+olm_bundle if {
+	input.image.config.Labels[olm_manifestv1]
+}
+
+_name(o) := n if {
+	n := o.name
+} else := "unnamed"
+
+# Finds all image references and their locations (paths)
+all_image_ref(manifest) := [e |
+	some imgs in [
+		[r |
+			some i, related in manifest.spec.relatedImages
```

Comment:
..or maybe some comments on each of these to help the reader understand each one.

Reply:
I think reference to the README above should be enough, I don't think the lookups are difficult to grasp, what they need to be is documented in that README. Would be nicer to point to a specification, but this is what we have...